### PR TITLE
[feat] creates the auth context

### PR DESCRIPTION
**Pull request by pragyakallanagoudar**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

- Adds the `AuthContext`, which has the (1) user session, (2) user ID, (3) user email, (4) sign in function, (5) sign out function, and (6) sign up function for use across the application.
- Removes the user ID and user email from `ProfileContext`, and instead modifies `ProfileContext` to retrieve user ID from the `AuthContext`.
- Refactors `/login`, `/signup`, and the sign out button functionality to use the `AuthContext` as opposed to calling the Supabase functions directly. 
- Refactors the components that originally used the user ID from the `ProfileContext` to use the `AuthContext` for this instead.
- **NOTE:** This does not refactor the `/reset-password` code to use the `AuthContext`, will discuss more tomorrow but I think we should probably look at the issues with auth emails first since we can't test reset password until that's resolved.

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @kyrenetam at the bottom."

N/A, all pages are the same visually

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

- Look at the newly created `src/utils/AuthProvider` and the changes to `src/utils/ProfileProvider`.
- Look at the changes to the auth pages/functionality on `/login`, `/signup`, and `/cases`. 
- Look at the changes in `src/components`.
- Run the app and test all auth-related functionality (from the previous two bullet points).
   - Before you do this, comment out anything in the `CaseDetails` component that uses legal experience -- this is no longer in the schema and so it errors.
- ❗**THINGS I'M CONFUSED ABOUT (help)**: 
   - Any modified dependencies for `useCallback`, `useMemo`, and `useEffect`
   - Anything marked with a TODO

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

- Refactor the `/reset-password` code to use the `AuthContext`.

## :link: Relevant Links

https://github.com/calblueprint/girls-write-now/blob/main/src/utils/AuthContext.tsx (huge adi ++)

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

N/A

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

#23 #37 (huge jinkang ++)

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @varortz

### CODE REVIEW

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: this seems a little redundant, we could just toss `newSession` into the function and handle the `null` case there (since it already handles falsy values)

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: can't we just toss `value.data.session`...

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: i wonder what error we'll have - if there's an error, should we still set these to be undefined?

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: should we remove this `console.log`?

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: i feel weird having `options` be an `object` type, since it won't provide us with any option suggestions in VS code when we use the function (as opposed to using `supabase.auth.signUp`.

It would be best if we can find a type that corresponds to what we want to pass into `supabase.auth.signUp`, either as a type import from supabase or as a custom defined interface

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx:119]: instead of this, i would recommend wrapping each function inside of a `useCallback` and just put the values into the `AuthContext.Provider` component (i.e. no `useMemo`).

reasoning (mansplaining):
- if **anything** is changed inside a React component, **everything** will be recomputed and re-rendered
- this means that if anything changes, all the functions (which are defined as local variables) will be re-rendered
- as a consequence, everything in the `useMemo` hook will be updated each time the component re-renders, which kind of defeats the whole purpose of the `useMemo` hook
- wrapping each function in a `useCallback` will fix this (and i recommend this), as it tells React to only update the function if anything in the dependency array of the `useCallback` hook is changed
    - in other words, the `useMemo` hook is no longer required
- an alternative is to define the functions inside of `useMemo` (as i have in the other contexts), but this is less efficient, since it will redefine **all** functions regardless of their respective dependencies each time **any** value updates

Comment (jinkang-0) [reply src/utils/AuthProvider.tsx:119]: i recognize i made the same mistake in `ProfileProvider` and `OnboardingProvider` as well LOL
would be great if you want to update that too 👍

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: i think error handling should be placed in their respective functions, no need for that here

File: src/utils/ProfileProvider.tsx
Comment (jinkang-0) [inline src/utils/ProfileProvider.tsx]: similar: ifl error handling should be handled in their respective functions

File: src/utils/ProfileProvider.tsx
Comment (jinkang-0) [inline src/utils/ProfileProvider.tsx]: response to `TODO`: yep auth should be a dependency, i would recommend adding the ESLint extension on VS Code

no issues here

File: src/app/(auth)/signup/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: we should probably remove this `console.log`

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: can't we just... pass in `options`...

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: flagging here as well: i think we can just toss `value.data.session`

File: src/components/CaseDetails/index.tsx
Comment (jinkang-0) [inline src/components/CaseDetails/index.tsx]: we can remove the `profile` check, since the values of the context is not used at all

additionally, can you replace the `profile` with `auth` in the dependency array?

File: src/app/settings/page.tsx
Comment (jinkang-0) [inline src/app/settings/page.tsx:168]: can you add `auth` to the dependency array of this `useCallback` hook?

File: src/app/settings/page.tsx
Comment (jinkang-0) [inline src/app/settings/page.tsx]: this is my bad, but we could lowk remove `React` from here, don't think we access that anywhere here lol

File: src/components/InterestForm/index.tsx
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: we're no longer using `useContext`, so we can remove that import

File: src/components/InterestForm/index.tsx
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: we don't need to use the `profile` context here, so we can remove this check

File: src/components/InterestForm/index.tsx
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: we won't need to use the profile context here, so we can delete this line

File: src/components/InterestForm/index.tsx
Comment (jinkang-0) [inline src/components/InterestForm/index.tsx]: since we won't be using the profile context, we don't need to import this hook

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: small nit: we can use `import type` for imports from `@supabase/supabase-js` and `crypto`, since we'll only need the type declarations

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: regarding the `TODO` - not sure how much it matters whether we're consistent with `null` or `undefined`, but I think there could be a subtle difference that can make things cleaner

for me, i see it as:
- `undefined` = the value haven't been initialized yet
- `null` = the value doesn't exist at all

ngl i think their difference is negligible and we can just use truthy value to differentiate those from usable info, but i'm curious to hear your thoughts on this

File: src/utils/OnboardingProvider.tsx
Comment (jinkang-0) [inline src/utils/OnboardingProvider.tsx]: we're no longer using `useContext`, so we can remove that import

File: src/utils/OnboardingProvider.tsx
Comment (jinkang-0) [inline src/utils/OnboardingProvider.tsx]: can you add `auth` to either:
- the dependency array of the `useMemo` hook; or
- if you decide to implement the `useCallback` hooks here, then the dependency array of that

File: src/utils/OnboardingProvider.tsx
Comment (jinkang-0) [inline src/utils/OnboardingProvider.tsx]: microscopic nit: should we standardize to call it `profile` vs `profileCtx`? not sure how much this matters. fine with either. 

side tip that im sure you know already: if we choose to refactor, you can put the text cursor on `profileCtx` then hit F2 when the keyword is highlighted to rename all instances of `profileCtx`

File: src/utils/OnboardingProvider.tsx
Comment (jinkang-0) [inline src/utils/OnboardingProvider.tsx]: i would say it's better to use `!auth.userId` here instead (idk what i was thinking before) since it would check for all of:
1. `''`
2. `null`
3. `undefined`

File: src/utils/ProfileProvider.tsx
Comment (jinkang-0) [inline src/utils/ProfileProvider.tsx]: i just realized we're checking for `auth.userId` but not using it here...
should we add a check for `auth.userId === newProfile.user_id`?

File: src/utils/OnboardingProvider.tsx
Comment (varortz) [reply src/utils/OnboardingProvider.tsx]: nit: I also prefer `profile` vs `profileCtx` lol

File: src/utils/AuthProvider.tsx
Comment (varortz) [reply src/utils/AuthProvider.tsx]: looked more into this and there are a lot of `null` haters, specifically for javascript/typescript. Also apparently Typescrip Coding Guidelines advise against the use of `null`, and Tony Hoare—the creator of null—considers it a billion dollar mistake 💀. Here are some articles I found:
- [We don't need null ](https://lou.cx/articles/we-don-t-need-null/)
- [null vs undefined? What to choose? What to use?](https://dev.to/sduduzog/null-vs-undefined-what-to-choose-what-to-use-11g)

so maybe `undefined` is the way to go for this use case?

Comment (jinkang-0) [reply src/utils/AuthProvider.tsx]: I don't think we should go full on `null` or full on `undefined`. They both have their own use cases, as the second article you provided described (thanks for doing the research I couldn't be bothered to do). I would be weary of following those practices to a T, especially if Typescript already gives us [a way to detect both at once](https://medium.com/@teamcode20233/understanding-double-question-mark-operator-in-typescript-be869f210fe6) (although this is more syntactic sugar).

I think, for the use case here, it makes sense for `signOut` to return `null` if there is no error. If we used `undefined`, it would sort of be saying that there was an error in getting the error, if that makes any sense at all. Or that the error wasn't set, even though it should've.

It also makes sense for `userId` and `userEmail` to be optionally `undefined`, because we may not have fetched that data yet. We can make an argument for `null` if the field can be fetched and is empty, but I don't think we'll encounter that error at least for `userId` and `userEmail`.

`session` is interesting though, since it can be `null` or `undefined`. In this case, I would go with defaulting to `null` just to have an empty value.

That being said, I do think we should be consistent across the app about our usage of `null` and `undefined`. I've probably made the same mistake about a hundred times.

If it makes it easier for everyone (since lowk I won't care what `null` or `undefined` means), we can go with `undefined` (borrowing [this philosophy](https://www.reddit.com/r/typescript/comments/11dpu05/comment/jaagvta/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)). This feels like taking the lesser evil to me though, and I still think there could be merit in encoding different semantic meanings in `undefined` and `null`.

Comment (jinkang-0) [reply src/utils/AuthProvider.tsx]: side note: we can reduce the definition to be `session?: Session`, which is Typescript's way of writing the same thing.

Comment (varortz) [reply src/utils/AuthProvider.tsx]: yeah, that makes sense. It does seem like there are cases where `null` makes sense over `undefined`, and we should probably try and be consistent across the app as you mentioned

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx]: is this the same concern as the one above? i.e. "we could just toss `newSession` into the function and handle the null case there" -- just want to make sure i'm catching everything

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx]: hmmm yeah i have no idea -- seems like the only one on [here](https://supabase.com/docs/reference/javascript/auth-signout) is you can't sign out if you're not already signed in, which shouldn't be possible anyways because of the way this is designed (i.e. sign out button doesn't show up if you're not logged in)

i think gwn does not check for the sign out error? not sure if we can just skip it

but yes im changing it so it only sets things to undefined if there is no error

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx]: are there situations where we will be calling `signUp` outside of the current instance in the signup page, though? this feels like a concern only if we were to call the function in different instances in the future, but it's not clear to me where this might be. maybe to make this cleaner, i can get rid of the `options` arg entirely since idk that it has much utility lol

File: src/utils/ProfileProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/ProfileProvider.tsx]: downloaded it, it better be life-altering

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx:119]: thanks jinkang!! i think i'm a little confused about the syntax of the fix lol so let's discuss on sunday

File: src/utils/OnboardingProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/OnboardingProvider.tsx]: i think the reason this is `profileCtx` here is because there is also this `useState` defined a few lines later:

`const [profile, setProfile] = useState<Partial<Profile>>({});`

shall i rename this instead? and if so, what to? maybe `onboardingProfile` or `profileObj` or `profileState` or smth?

File: src/components/CaseDetails/index.tsx
Comment (pragyakallanagoudar) [reply src/components/CaseDetails/index.tsx]: hmm can you clarify how they're not being used? aren't they relevant for the un-onboarded state

File: src/components/InterestForm/index.tsx
Comment (pragyakallanagoudar) [reply src/components/InterestForm/index.tsx]: for this and all the other comments about getting rid of the `ProfileContext` in the `InterestForm` -

@varortz I was thinking the `ProfileContext` would be relevant to the profile match -- should I leave this here for that reason?

File: src/utils/OnboardingProvider.tsx
Comment (jinkang-0) [reply src/utils/OnboardingProvider.tsx]: oh mb i overlooked that

maybe we can call it `userProfile`?

File: src/components/CaseDetails/index.tsx
Comment (jinkang-0) [reply src/components/CaseDetails/index.tsx]: oh i overlooked the context.

currently you're only checking if the `profile` object exists, which represents the profile context. But that doesn't tell us anything about the actual onboarded-ness of the user.

taking into account the context of the use case, a better suggestion might be to use `profile.profileData != null` instead

File: src/utils/ProfileProvider.tsx
Comment (jinkang-0) [reply src/utils/ProfileProvider.tsx]: it'll be just like brookies

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [reply src/utils/AuthProvider.tsx]: yes

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [reply src/utils/AuthProvider.tsx]: ohh that makes sense

on a slightly less relevant tangent --
looking at the docs, i noticed this blurb:
> By default, `signOut()` uses the global scope, which signs out all other sessions that the user is logged into as well.

given this, should we add a check in `onAuthStateChange` to set those fields to be `undefined` when we receive a log out event? In fact, should we just move all our state changes into the `onAuthStateChange` function depending on the event received?

not sure if that would work, but it may make things a lot cleaner and secure

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [reply src/utils/AuthProvider.tsx]: I doubt it, but it's still good practice to have well-defined types. I can imagine a scenario where we may want to modify arguments to the `signUp` function but get mildly inconvenienced by the lack of IntelliSense. Which is the worst inconvenience to have if I'm being completely honest.

Anywho, if it helps, a quick dive into the node modules give us this declaration:
```Typescript
options?: {
  emailRedirectTo?: string;
  data?: object;
  captchaToken?: string;
}
```

But I'll leave the choice (and blame) of implementing this up to you.

note that `options` is optional tho (can be `undefined`)

File: src/components/InterestForm/index.tsx
Comment (varortz) [reply src/components/InterestForm/index.tsx]: I think profile match would be its own separate component, so it should be ok to remove it in the `InterestForm` if we're not using it

File: src/utils/OnboardingProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/OnboardingProvider.tsx]: hmmm `userProfile` (and `userState`, my other idea) also don't work because there are types imported from `@/types/schema` that have the same name. imma use `profileObj` lol

Comment (jinkang-0) [reply src/utils/OnboardingProvider.tsx]: Can you elaborate on where they're imported? I don't see `userProfile` or `userState` being defined (nor exported) from `@/types/schema`
image
image

Comment (jinkang-0) [reply src/utils/OnboardingProvider.tsx]: but `profileObj` would work too

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [reply src/utils/AuthProvider.tsx:109]: In hindsight, I doubt we'd need to use a `useCallback` hook at all, since they don't return a value for the memoization to be helpful. We could probably just define the functions as we would for normal handlers and toss that along the other values (which may or may not be wrapped in `useMemo`).

File: src/utils/OnboardingProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/OnboardingProvider.tsx]: i actually. have literally no clue what i was referring to LOL. `userProfile` worked! replacing

File: src/utils/OnboardingProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/OnboardingProvider.tsx]: added it to the dependency array of the useMemo hook!

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx:119]: (Recap of what we discussed this during MPH) It seems like using this causes a TS compiler error, so I'm just using what we already had for right now. but @jinkang-0 @varortz curious if you have other approaches you'd like to suggest?

Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx:119]: Update: I wrapped the auth functions in `useCallback`s and left everything else as is, and this seems to get rid of all TS compiler errors -- but let me know what y'all think!!

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx]: Sounds good! I've replaced it with `session?: Session`

File: src/utils/AuthProvider.tsx
Comment (pragyakallanagoudar) [reply src/utils/AuthProvider.tsx]: @varortz thoughts on moving things to `onAuthStateChange`? im chill w anything

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: nit: we can wrap this in a `useCallback` as well

File: src/utils/OnboardingProvider.tsx
Comment (jinkang-0) [inline src/utils/OnboardingProvider.tsx]: for this and all other functions in the `useMemo`, could you put them outside the `useMemo` and wrap each of them individually in a `useCallback`?

with the current configuration, all of these functions will be re-defined every time the `useMemo` is re-computed (which happens a lot) - ideally we'd just want to define functions once.

File: src/utils/ProfileProvider.tsx
Comment (jinkang-0) [inline src/utils/ProfileProvider.tsx]: nit: correct me if I'm wrong, but since we check for `auth?.userId` (which checks `auth` and `auth.userId`), I think we can just drop the question marks here

File: src/utils/ProfileProvider.tsx
Comment (jinkang-0) [inline src/utils/ProfileProvider.tsx]: can you do something similar with the `useCallback`s here? this is completely my fault - sorry about this.

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: nit: we can use the optional marker here (i.e. `userId?: UUID`)

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx:24]: nit: let's use `SignUpOptions` here

Comment (jinkang-0) [reply src/utils/AuthProvider.tsx:24]: also thoughts on consolidating these miscellaneous types into a `@/types/misc.d.ts` file? (cc @varortz )

**Review by jinkang-0 — APPROVED** [review]
amazing work pragya! should be good to go after rebasing and iterations, I'll approve now to speed things up a bit, but feel free to ping me if you want me to take another look

File: src/utils/AuthProvider.tsx
Comment (varortz) [reply src/utils/AuthProvider.tsx:24]: would this also include types like `OnboardingContextType`, `ProfileContextType`?

Comment (jinkang-0) [reply src/utils/AuthProvider.tsx:24]: No, I think context types are best left in the relevant context files. In hindsight, it may be better to leave the `SignUpOptions` type here since we don't need to reference the type anywhere else as well.

Comment (varortz) [reply src/utils/AuthProvider.tsx:24]: Yeah, if it isn't used somewhere else, it's probably best to leave it in this file.

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: can we remove this console log?

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: can we remove this console log?

**Review by varortz — APPROVED** [review]
looks great!! Thank u so much for all ur hard work :))

**Pull request merged**
